math/src/elliptic_curve/short_weierstrass/curves/bls12_381/twist.rs

```diff
@@ -46,7 +47,72 @@ impl IsShortWeierstrass for BLS12381TwistCurve {
     }
 }
 
+impl BLS12381TwistCurve {
+    pub fn is_on_curve(point: &<Self as IsEllipticCurve>::PointRepresentation) -> bool {
+        let [x, y, z] = point.coordinates();
+
+        let lhs = {
+            let y_sq = <Self as IsEllipticCurve>::BaseField::square(y.value());
+            <Self as IsEllipticCurve>::BaseField::mul(&y_sq, z.value())
+        };
+
+        let rhs = {
+            let x_sq = <Self as IsEllipticCurve>::BaseField::square(x.value());
+            let x_cubed = <Self as IsEllipticCurve>::BaseField::mul(&x_sq, x.value());
+
+            let z_sq = <Self as IsEllipticCurve>::BaseField::square(z.value());
+            let z_cubed = <Self as IsEllipticCurve>::BaseField::mul(&z_sq, z.value());
+
+            let z_cubed_with_b =
+                <Self as IsEllipticCurve>::BaseField::mul(&z_cubed, Self::b().value());
+
+            <Self as IsEllipticCurve>::BaseField::add(&x_cubed, &z_cubed_with_b)
+        };
+
+        lhs == rhs
+    }
+}
```

Comment:
Something along this lines should do the work. Slower performancewise probably. But I wouldn't worry about that, since we didn't tackle optimizations of elliptic curve operations yet.
```suggestion
    pub fn is_on_curve(point: &<Self as IsEllipticCurve>::PointRepresentation) -> bool {
        if point.z() == &FieldElement::zero() {
            true
        } else {
            let point = point.to_affine();
            Self::defining_equation(&point.x(), &point.y()) == FieldElement::zero()
        }
    }
```
On the other hand. I saw that you use this method on a point that's already of type `ShortWeierstrassProjectivePoint<BLS12381TwistCurve>`. That shouldn't be necessary. If the current state of the code allows one to instantiate an element of that type that's not on the curve, then that's a bigger problem. If you suspect that's the case, please feel free to create an issue.

Reply:
Can you explain it more?

Reply:
`ShortWeierstrassProjectivePoint<BLS12381TwistCurve>` is supposed to model a point on the twist. If there is an instance of it for which the defining equation of the twist does not hold, then there's a constructor failing to raise a creation error.